source/s3_client.c

```diff
@@ -257,12 +279,43 @@ struct aws_s3_client *aws_s3_client_new(
     struct aws_s3_client *client = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_client));
 
     client->allocator = allocator;
+
+    size_t mem_limit = 0;
+    if (client_config->memory_limit_in_bytes == 0) {
+#if SIZE_BITS == 32
+        if (client_config->throughput_target_gbps > 25.0) {
+            mem_limit = GB_TO_BYTES(2);
+        } else {
+            mem_limit = GB_TO_BYTES(1);
+        }
+#else
+        if (client_config->throughput_target_gbps > 75.0) {
+            mem_limit = GB_TO_BYTES(8);
+        } else if (client_config->throughput_target_gbps > 25.0) {
+            mem_limit = GB_TO_BYTES(4);
+        } else {
+            mem_limit = GB_TO_BYTES(2);
+        }
+#endif
+    } else {
+        mem_limit = client_config->memory_limit_in_bytes;
+    }
+
+    size_t part_size;
+    if (client_config->part_size != 0) {
+        part_size = (size_t)client_config->part_size;
+    } else {
+        part_size = s_default_part_size;
+    }
+
+    client->buffer_pool = aws_s3_buffer_pool_new(allocator, part_size, mem_limit);
```

Comment:
need to check that this succeeded. It will raise an error if mem_limit isn't just right